libraries/lib-concurrency/ParallelReduce.h

```diff
@@ -0,0 +1,307 @@
+ `report_progress` which accepts the total progress value and is called only on
+ the original task.
+ */
+template<typename Task, typename Increment,
+   MergeChoice doMerge, ReportChoice doReport>
+struct TaskAdapter : Task {
+   using Task::Task;
+   TaskAdapter(
+      TaskAdapter &orig, typename Task::iterator_type iter)
+      : Task{ orig, iter }
+      , is_first{ false }
+   {}
+   void run(Progress<Increment> &progress)
+   {
+      std::for_each(this->begin(), this->end(), [this, &progress](auto &&value){
+         if constexpr (doReport) {
+            auto total =
+               progress((*this)(std::forward<decltype(value)>(value)));
+            if (is_first)
+               this->report_progress(total);
+         }
+         else {
+            (*this)(std::forward<decltype(value)>(value));
+            progress(Increment{});
+         }
+      });
+   }
+   void merge(TaskAdapter &&other, Progress<Increment> &progress) {
+      if constexpr (doMerge) {
+         if constexpr (doReport) {
+            auto total = progress(this->Task::merge(std::move(other)));
+            if (is_first)
+               this->report_progress(total);
+         }
+         else {
+            this->Task::merge(std::move(other));
+            progress(Increment{});
+         }
+      }
+      else
+         progress(Increment{});
+   }
+private:
+   bool is_first{ true };
+};
+
+namespace detail {
+template<typename Iterator, typename Callable>
+struct ForEach_Task : SplittableBase<Iterator> {
+   using iterator_type = Iterator;
+
+   ForEach_Task(Iterator first, Iterator last, Callable callable)
+      : SplittableBase<Iterator>{ first, last }
+      , callable{ std::move(callable) }
+   {}
+   ForEach_Task(ForEach_Task &task, Iterator iterator)
+      : SplittableBase<Iterator>{ task, iterator }
+      , callable{ task.callable }
+   {}
+
+   template<typename Value> void operator ()(Value &&value) {
+      return callable(std::forward<Value>(value)); }
+
+   Callable callable;
+};
+}
+
+//! A convenient type for a body of a simple parallelized for_each
+template<typename Iterator, typename Callable> struct ForEach
+: TaskAdapter<detail::ForEach_Task<Iterator, Callable>,
+   typename std::iterator_traits<Iterator>::difference_type,
+   NoMerge, NoReport
+>{
+   using TaskAdapter<detail::ForEach_Task<Iterator, Callable>,
+      typename std::iterator_traits<Iterator>::difference_type,
+      NoMerge, NoReport
+   >::TaskAdapter;
+};
+
+// Deduction guide
+template<typename Iterator, typename Callable>
+ForEach(Iterator, Iterator, Callable) -> ForEach<Iterator, Callable>;
+
+template<typename Task, typename Increment>
+void OneDimensionalReduce(Task &task, unsigned n_threads)
+{
+   assert(n_threads >= 1);
+   // TODO: This function might be modified instead to allocate tasks to threads
+   // from a pool, and the Task class might define other policies for
+   // subdivision of its range
+   n_threads = std::max(1u, n_threads);
```

Comment:
I like that the assertion is not the only check, but do we need the `n_threads > 0` at all?